lib/fluent/plugin/out_google_cloud.rb

```diff
@@ -612,14 +620,8 @@ def write(chunk)
                                             ts_secs,
                                             ts_nanos)
 
-          # Get fully-qualified trace id for LogEntry "trace" field.
-          fq_trace_id = record.delete(@trace_key)
```

Comment:
This breaks the use case where the user's JSON contains those fields instead.

Reply:
These keys (https://github.com/GoogleCloudPlatform/fluent-plugin-google-cloud/blob/8d26465c0f777a42938e3acae857936627a38511/lib/fluent/plugin/out_google_cloud.rb#L149) are reserved for LogEntry fields. By default they come with the `logging.googleapis.com/` prefixes. It should not conflict with other fields in users' namespace if that was the concern.

On the other hand, before this PR we already had the logic (https://github.com/GoogleCloudPlatform/fluent-plugin-google-cloud/blob/8d26465c0f777a42938e3acae857936627a38511/lib/fluent/plugin/out_google_cloud.rb#L617) to delete these fields from the record.

Reply:
You're missing the fact that when the text payload is detected to be serialized JSON, the parsed JSON becomes the record. So users previously could serialize JSON with those fields into text, which would then be extracted into the log entry. This would no longer be possible after this change as-is.
If you want to make the behavior consistent, you should remove it from the record right before detecting JSON, and then reinstate it in the resulting record only if it's not already present (e.g., the user's code dumped its own value of insert_id into the JSON).

That said, what tools would set the trace_id or span_id in a record that also contains serialized JSON?

Reply:
Oh, I think I misunderstood. You were talking about the LogEntry fields being set in the nested JSON. Then maybe we should check both levels.

Reply:
Oops, race condition. Did not see the second comment before I commented.

How about something like this? https://github.com/GoogleCloudPlatform/fluent-plugin-google-cloud/pull/247/commits/34effd81aa1433e972ebe8105444f06c33c17412

Is it worth checking these fields on both levels?

I'm not sure if there are tools that would set `trace_id` or `span_id` in a record that also contains serialized JSON. If they use the `record_transformer` plugin to reformat it somehow, it might end up being at the root level.